packages/SwingSet/src/vats/comms/clist-inbound.js

```diff
@@ -29,46 +19,43 @@ export function makeInbound(state, stateKit) {
 
   function retireRemotePromiseID(remoteID, rpid) {
     insistRemoteType('promise', rpid);
-    const remote = getRemote(state, remoteID);
-    const lpid = remote.fromRemote.get(rpid);
+    const remote = state.getRemote(remoteID);
+    const lpid = remote.mapFromRemote(rpid);
     assert(lpid, X`unknown remote ${remoteID} promise ${rpid}`);
-    const p = state.promiseTable.get(lpid);
+    const p = state.getPromise(lpid);
     assert(
       !p.subscribers || p.subscribers.indexOf(remoteID) === -1,
       X`attempt to retire remote ${remoteID} subscribed promise ${rpid}`,
     );
-    remote.fromRemote.delete(rpid);
-    remote.toRemote.delete(lpid);
+    remote.deleteRemoteMapping(rpid, lpid);
     cdebug(`comms delete mapping r<->k ${remoteID} {rpid}<=>${lpid}`);
   }
 
   function beginRemotePromiseIDRetirement(remoteID, rpid) {
     insistRemoteType('promise', rpid);
-    const remote = getRemote(state, remoteID);
-    const lpid = remote.fromRemote.get(flipRemoteSlot(rpid));
-    remote.toRemote.delete(lpid);
-    remote.retirementQueue.push([remote.nextSendSeqNum, rpid]);
+    const remote = state.getRemote(remoteID);
+    const lpid = remote.mapFromRemote(flipRemoteSlot(rpid));
+    remote.deleteToRemoteMapping(lpid);
+    remote.enqueueRetirement([remote.nextSendSeqNum(), rpid]);
     cdebug(`comms begin retiring ${remoteID} ${rpid} ${lpid}`);
   }
 
   function retireAcknowledgedRemotePromiseIDs(remoteID, ackSeqNum) {
-    const remote = getRemote(state, remoteID);
-    while (remote.retirementQueue.length > 0) {
-      const [sentSeqNum, rpid] = remote.retirementQueue[0];
-      if (sentSeqNum <= ackSeqNum) {
+    const remote = state.getRemote(remoteID);
+    for (;;) {
+      const rpid = remote.nextReadyRetirement(ackSeqNum);
```

Comment:
I don't know how `nextReadyRetirement` works, I haven't gotten that far yet, but if it keeps the entire ready-to-retire list in a single serialized Array (which is what I would do, it ought to be short as long as the two sides are still talking to each other), then this would be more efficient to do the loop in the place that has the whole list. Deserialize the whole thing into an Array, loop and `shift()`, then re-serialize it back to disk at the end. A split like this is likely to need to deserialize+reserialize the list once per retired promise.

Reply:
That's a little awkward, since the "how to retire" logic is in `clist-inbound.js` whereas `remote.js` only knows about holding state.  But I did the moral equivalent: I replaced `nextReadyRetirement` with a `getReadyRetirements` that swoops through the queue in one pass and pulls out all the ready elements, only doing one deserialization/update/reserialization for the whole thing.  Then I changed `retireAcknowledgedRemotePromiseIDs` to loop over the ready retirements.  This actually made both functions smaller :-)

Reply:
Ok, so that adds a responsiblity to the caller of `getReadyRetirements()`, right? It's no longer idempotent, and if the caller doesn't actually retire them, they'll be lost forever? That's a fine design, but maybe it deserves a comment or a function name that implies the side effect / state change. `removeReadyRetirements()`???

Reply:
That's what it did before; it wasn't idempotent then either, it just did them one at a time.  Changing the name from `get...` to `remove...` would make the responsibility clearer, though.